### ISSUE TITLE
OutOfMemory error due to high memory usage of io.aleph.dirigiste.Pool

### ISSUE THREAD

**Issue by rborer** [ztellman/aleph#394]
Hi,

I'm currently incorporating Aleph into https://paper.li/ service as a replacement for our venerable crawler (based on Netty 3.10!).

On our test environment (which is crawling around 350 different URLs per minute, much less than our production system), my crawling component is crashing in OutOfMemory after a few hours of processing 😢 .

I started analysing the heap dumps and it seems to indicate that `io.aleph.dirigiste.Pool` is retaining the majority of heap as shown below (Xmx is set to 9G):

<img width="835" alt="screen shot 2018-06-25 at 09 40 42" src="https://user-images.githubusercontent.com/361403/41836781-04880c58-785c-11e8-9230-fd4fd113bc00.png">

From my understanding so far, this high memory usage seems to be related to the Pool statistics. As shown below, a single `java.util.HashMap` (for the statistics probably) is retaining 40% of the heap:

<img width="710" alt="screen shot 2018-06-25 at 09 43 55" src="https://user-images.githubusercontent.com/361403/41836937-77ba9e3e-785c-11e8-9609-21f9ddf46e9e.png">

And memory is mostly filled with longs & doubles that seems related to the Pool statistics:

<img width="519" alt="screen shot 2018-06-25 at 09 44 03" src="https://user-images.githubusercontent.com/361403/41837089-cd45318e-785c-11e8-82fe-462169577fb3.png">

Am I correct in my analysis to finger pointing aleph Pool? If yes, is there any setting on my side I can modify to relieve this memory usage?

Thanks.

**Comment by gsnewmark** [comment]
May be related https://github.com/ztellman/dirigiste/pull/5#issuecomment-159140122

**Comment by rborer** [comment]
I was finally able to run [Eclipse Memory Analyzer](https://www.eclipse.org/mat/) on my laptop (previously I hit a bug between OSX, SWT & Eclipse that would freeze it at startup).

So I could drill-down into the `Pool` instance, to see that indeed it is the `_stats` field that is the culprit:

<img width="1281" alt="screen shot 2018-06-28 at 10 52 15" src="https://user-images.githubusercontent.com/361403/42023956-63a99578-7ac1-11e8-92c1-749fe223b8d7.png">

Here we can see that this HashMap instance contains 29k entries (not much), but obviously each entry consumes more than 100mb of retained heap!

This strongly advise that https://github.com/ztellman/dirigiste/pull/18 is going into the right direction. Although stats are memory heavy, this Dirigiste patch will ensure that they are not kept in memory when they are not needed anymore!

**Comment by rborer** [comment]
I've been running aleph with https://github.com/ztellman/dirigiste/pull/18 applied. Although it doesn't prevent my process to OOM, it seems to be able to run for a longer time, so there is progress with this patch!

the heap dump now show that this HashMap instance is not present in the heap anymore (thanks to the JVM escape analysis it's on the stack now 😉).

The heap dump analysis how shows to new culprits: UniformLongReservoirMap and UniformDoubleReservoirMap:

<img width="714" alt="screen shot 2018-07-12 at 07 28 29" src="https://user-images.githubusercontent.com/361403/42614878-0683d9e8-85a8-11e8-9cb3-8a4bccb9d735.png">

So, my naive question is: can we decrease the [`RESERVOIR_SIZE`](https://github.com/ztellman/dirigiste/blob/a71ee4dc3a97e79c7c5ac5c58d465ab8e6bf3519/src/io/aleph/dirigiste/Stats.java#L25) which is set to retain 4'096 longs?

**Comment by ztellman** [comment]
We could, though that would make p999 statistics from the pool fairly useless.  I suspect, however, that what's happening here is that the number of these reservoirs is growing monotonically as you crawl over new domains, rather than disappearing once you exhaust all the pages.  I apologize for the delay in looking at this, things have been a bit hectic on my end.  I'll try to carve out some time before the week's end.

**Comment by rborer** [comment]
I've tried to address this ever increasing memory needs under https://github.com/ztellman/dirigiste/pull/19 , feedback is welcomed :-)

**Comment by rborer** [comment]
@ztellman I can confirm that the dirigiste patch https://github.com/ztellman/dirigiste/pull/19 has made the OOM an issue of the past. It's been running for multiple days with a stable memory use 😄.

Did you had time to review this PR already?

**Comment by danielcompton** [comment]
@rborer I think Zach is on holiday at the moment.

**Comment by rborer** [comment]
Thanks for the dirigiste release @ztellman (and sorry if I annoyed you during your holidays 😬).

I've been running since a few weeks with dirigiste 0.1.6-alpha1 successfully. Although I still have some infrequent OOM, at first glance they don't seems related to dirigiste anymore so from my point of view the issue is fixed.

Maybe you want to ship a new aleph alpha release with this included to have broader testing by the community?

**Comment by ztellman** [comment]
This is included in the latest alpha.

**Comment by rborer** [comment]
@ztellman actually I don't think it's included within aleph 0.4.7-alph2 :thinking: . Here is the dependency tree I see:

```
 [aleph "0.4.7-alpha2"]
   [byte-streams "0.2.5-alpha2"]
     [clj-tuple "0.2.2"]
     [primitive-math "0.1.6"]
   [io.netty/netty-codec-http "4.1.30.Final"]
   [io.netty/netty-codec "4.1.30.Final"]
   [io.netty/netty-handler-proxy "4.1.30.Final"]
     [io.netty/netty-codec-socks "4.1.30.Final"]
   [io.netty/netty-handler "4.1.30.Final"]
   [io.netty/netty-resolver-dns "4.1.30.Final"]
     [io.netty/netty-codec-dns "4.1.30.Final"]
   [io.netty/netty-resolver "4.1.30.Final"]
   [io.netty/netty-transport-native-epoll "4.1.30.Final"]
     [io.netty/netty-common "4.1.30.Final"]
     [io.netty/netty-transport-native-unix-common "4.1.30.Final"]
   [io.netty/netty-transport "4.1.30.Final"]
     [io.netty/netty-buffer "4.1.30.Final"]
   [manifold "0.1.8"]
     [io.aleph/dirigiste "0.1.5"]
     [riddley "0.1.14"]
   [potemkin "0.4.5"]
```

`dirigiste` is still transitively set at version "0.1.5" while the fix is included into version "0.1.6-alpha1".

Am I missing something?

**Comment by ztellman** [comment]
Okay, this is definitely now included in `0.4.7-alpha3`.

**Comment by rborer** [comment]
Thanks for the release :smiley:
